activerecord/lib/active_record/transactions.rb

```diff
@@ -209,7 +209,9 @@ module Transactions
     module ClassMethods
       # See the ConnectionAdapters::DatabaseStatements#transaction API docs.
       def transaction(**options, &block)
-        connection.transaction(**options, &block)
+        with_connection do |connection|
+          connection.transaction(**options, &block)
+        end
```

Comment:
This seems like a fairly significant breaking change.

Speaking from my role as GoodJob maintainer, which relies on connection/session-based Advisory Locks, I think this is a big deviation from Active Record behavior of connection fixation. **This breaks GoodJob a lot.** I feel like I've just made progress on making sure people use a Rails Executor to not end up with unexpected connection behavior, and this now adds an additional layer to that. 

This written, I do _like_ the overall direction of being explicit about connection-fixation and allowing per-transaction/statement leasing. But maybe more configurably opt-in-able. One idea is having this behavior be a property of the Rails Executor.

Reply:
I was helped by @matthewd to understand that I won't have a problem so long as I use a workaround to access the `connection` before entering a transaction.

Reply:
Yes, this doesn't change the previous `#connection` behavior. It will be changed eventually, thought.

In the case of Good Job, you are probably best to just lease it for the duration of the job anyway.

Reply:
I think this reveals an existing issue with `connection`: if someone calls it directly, we should pin that connection for the rest of the request, because they might have assumed it will persist (either by retaining a reference, or by manipulating session-level state).

But right now, `with_connection` (and now, by extension and more visibly, `with_transaction`) will release a connection [that it acquired] back to the pool, _even if the user code in the inner block called `connection`_.

On reflection, I think we need to perma-pin the connection in that case, at the time `connection` is called, despite the wrapping `with_connection` block originally believing it should give it back.

Reply:
Right, I was hoping it wouldn't be necessary, but I'm afraid it is. I can try to fix that.

Reply:
Yeah.. while some of @bensheldon's use case is clearly non-standard, I think it is pointing to a general risk, which is that the [here inside-transaction] code wrapped by the block is not necessarily aware of what's going on further up the stack.

At least as we transition internals to `with_connection` it won't affect us -- and assuming we reach a point where `connection` is _only_ for user access, it serves as a reasonable signpost for us to switch into safety mode.

Definitely unfortunate that it means we need to keep a second bit of data per thread, though. I guess we could use some sort of transient/permanent flag on the connection's lease, rather than keeping it separately? 🤷🏻‍♂️

Reply:
Yeah, I have an idea how to handle that.

Reply:
I don't think it's strictly necessary to fixate a connection that is checked out _within_ `with_connection` after that block closes, at least for my needs. It was just not understanding that there was a pre-existing contract that I needed to checkout the connection [_before_](https://github.com/bensheldon/good_job/pull/1259/files#diff-90c5de68d2e35022573af8a54ce6b45b741a21ecf9845a7e66674c1ed3c91d0eR399-R400) entering `with_connection`. 

> I think it is pointing to a general risk, which is that the [here inside-transaction] code wrapped by the block is not necessarily aware of what's going on further up the stack.

fyi, I'll share that this is the challenge of building Rails Plugins/Engines (that build on top of Rails libraries): these style of blocks (like Rails Executor) that clear/reset state that get inserted into call stacks that are like:

```ruby 
framework do
  application do
    plugin do
      framework # <-- when new state resets get inserted
    end # <-- then behavior here can break
  end
end
```